### PR TITLE
Add nondeterministic alert for `MaxUnpool1d/2d/3d`

### PR DESCRIPTION
**Pull request by kurtamohler**
Part of #80827
Part of #78249

### CODE REVIEW

**Review by lezcano — APPROVED** [review]
LGTM, but let's wait for @nikitaved see if he has anything to add.

File: aten/src/ATen/native/MaxUnpooling.cpp [hunk 1]
```diff
@@ -11,6 +11,10 @@ Tensor& max_unpooling2d_forward_out_cpu(
     const Tensor& indices_,
     IntArrayRef output_size,
     Tensor& output) {
+  // See Note [Writing Nondeterministic Operations]
+  // Nondeterministic with duplicate indices
+  at::globalContext().alertNotDeterministic("max_unpooling2d_forward_out");
```
Comment (nikitaved) [inline aten/src/ATen/native/MaxUnpooling.cpp:16]: Am I missing something? But shouldn't we have these alerts enabled for 1D case as well?

File: test/test_torch.py [hunk 2]
```diff
@@ -1242,6 +1242,45 @@ def backward_func(slf, device):
...
 
+    @dtypes(*floating_types())
+    @onlyNativeDeviceTypes
+    def test_nondeterministic_alert_MaxUnpool1d(self, device, dtype):
+        module = torch.nn.MaxUnpool1d(3, 1)
+        input = torch.randn(1, 1, 7, dtype=dtype, device=device)
+        indices = torch.zeros_like(input, dtype=torch.long, device=device)
+
+        @expectedAlertNondeterministic('max_unpooling2d_forward_out')
+        def forward_func(slf, device):
```
Comment (nikitaved) [inline test/test_torch.py:1253]: the 1D case uses 2D under the hood?

File: test/test_torch.py [hunk 3]
```diff
@@ -1242,6 +1242,45 @@ def backward_func(slf, device):
 
         backward_func(self, device)
 
+    @dtypes(*floating_types())
+    @onlyNativeDeviceTypes
```
Comment (nikitaved) [inline test/test_torch.py:1246]: Does `maxunpool` work for dtypes other than floats? Maybe we should test them as well...

File: test/test_torch.py [hunk 4]
```diff
@@ -1242,6 +1242,45 @@ def backward_func(slf, device):
...
+    def test_nondeterministic_alert_MaxUnpool3d(self, device, dtype):
+        module = torch.nn.MaxUnpool3d(3, 1)
+        input = torch.randn(1, 1, 7, 7, 7, dtype=dtype, device=device)
+        indices = torch.zeros_like(input, dtype=torch.long, device=device)
+
+        @expectedAlertNondeterministic('max_unpooling3d_forward_out')
+        def forward_func(slf, device):
+            module(input, indices)
+
+        forward_func(self, device)
```
Comment (nikitaved) [inline test/test_torch.py:1282]: Just curious. Does `forward_func` use `self/device` at all?

Comment (kurtamohler) [reply test/test_torch.py:1282]: No, but they're required for the decorator. #84808 will make these tests easier to read

File: test/test_torch.py (hunk 3 above)
Comment (kurtamohler) [reply test/test_torch.py:1246]: I tried complex, half, and ints, and they didn't seem to be supported

File: test/test_torch.py (hunk 2 above)
Comment (kurtamohler) [reply test/test_torch.py:1253]: That's right

File: torch/__init__.py [hunk 5]
```diff
@@ -416,6 +416,9 @@ def use_deterministic_algorithms(mode, *, warn_only=False):
         * :class:`torch.nn.AdaptiveMaxPool2d` when attempting to differentiate a CUDA tensor
         * :class:`torch.nn.FractionalMaxPool2d` when attempting to differentiate a CUDA tensor
         * :class:`torch.nn.FractionalMaxPool3d` when attempting to differentiate a CUDA tensor
+        * :class:`torch.nn.MaxUnpool1d`
```
Comment (kurtamohler) [inline torch/__init__.py:419]: @mruberry, in response to your comment here: https://github.com/pytorch/pytorch/pull/78838#discussion_r971959851

> Shouldn't the warning also be on the functional variants? I think these modules just call their functional counterparts

You're right, the modules do just call their functional counterparts, and that is actually where the alerts are being generated from (I'm almost sure that that's true of all of the modules listed here, not just MaxUnpool). I should add tests for all of the functional variants and add them all to the documentation as well.

**Review by mruberry — APPROVED** [review]
Stamping -- although as @Lezcano says this may want to wait for @nikitaved's review

File: test/test_torch.py (hunk 3 above)
Comment (kurtamohler) [reply test/test_torch.py:1246]: Nope, looks like float and double are the only supported dtypes for both devices. Half is supported by one of the devices, but not the other, so I could add that in if we really want

Comment (nikitaved) [reply test/test_torch.py:1239]: Let's add it just in case as well? Half is used in the context of quantized nns.

**Review by nikitaved — APPROVED** [review]
LGTM! Thank you, Kurt! Let's just probably add that Half test :)

**Comment by kurtamohler** [comment]
I think that merge failure looks flaky, trying again

**Pull request closed without merging**